keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -0,0 +1,953 @@
+strategic merge patch or apply) will be less performant than not doing so.
+[Initial
+benchmarks](https://github.com/kubernetes/kubernetes/pull/104433#issuecomment-901398507) estimate ~20% slower 25-30% more memory consumption.
+
+This is deemed acceptable and insignificant because the use case we are
+replacing is client side validation that happens from kubectl (i.e. via a human
+that should not be too impacted by a slight performance hit). Any existing
+automated clients that would be impacted by performance do not currently have a
+way of leveraging server side validation and thus will remain opted-out of this
+by default (or will be willing to tradeoff the performance if they do choose to
+opt-in to server side schema validation).
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### Opt-in API Mechanism
+
+There are a few ways we could allow a client to opt-in to server-side unknown
+field validation, we present two options **Content-Type Header** and **query
+parameter**.
+
+#### Content-Type Header
+
+Requests that contain a kubernetes object, also pass along with it a
+content-type header such as “application/json”. One way to indicate to the
+server that it should use strict schema validation and fail when unknown fields
+are passed is to send a mime-type parameter indicating the strict validation
+such as “application/json;validation=strict”. Alternatively we could use a new
+header such as “X-Kubernetes-Validation:strict”.
+
+One could argue that this is the more appropriate way to parameterize opting-in
+to server side schema validation because on the server we are fundamentally
+treating the content as a different type (“application/json” is json data that
+we don’t care if it has extra fields, “application/json;validation=strict” is
+json data that is sensitive to extra fields).
+
+A precedent for using a mime-type parameter is from how we [receive resources as
+tables](https://kubernetes.io/docs/reference/using-api/_print/#receiving-resources-as-tables).
+
+On the other hand, one could also argue that interpreting input strictly or not
+according to the target schema is independent of the content type and that this
+is an inappropriate way to parameterize validation opt-in.
+
+Another argument against this method is that for patch, strictness is handled
+when decoding the *result* of the patch, so it wouldn’t make sense to use
+Content-Type which is providing information about the inbound request body.
+
+#### Query Parameter
+
+Alternatively, if we don’t like the idea of using Content-Type header to
+determine whether the apiserver should accept or fail on unknown fields, we
+could pass a query param such as “?fieldValidation=Strict”.
```

Comment:
Will I have the option of `fieldValidation=Warn`?

Reply:
I've added to note to this section indicating that we could support such an option. I'm open to it, but I don't think it's necessary for alpha. I think it raises further questions around whether "Warn" or "Ignore" should be the default option (and the performance implications of having "Warn" be the default option) that I think are best left out of the alpha stage.

Let me know if you disagree and want us to commit to having a "Warn" option here in alpha though, because I don't feel too strongly either way.

Reply:
> Let me know if you disagree and want us to commit to having a "Warn" option here in alpha though, because I don't feel too strongly either way.

I would like to have a Warn option.

Reply:
Sounds good, I've modified the note to commit to support the "Warn" the option.